File: branches/convenience-views/implementation/UrakawaToolkit/metadata/IMetadataFactory.cs
using System;

namespace urakawa.metadata
{
	/// <summary>
	/// Interface of factory for constructing <see cref="IMetadata"/>
	/// </summary>
	public interface IMetadataFactory
	{
    /// <summary>
    /// Creates an <see cref="IMetadata"/> matching a given QName
    /// </summary>
		/// <param name="localName">The local part of the QName</param>
		/// <param name="namespaceUri">The namespace uri part of the QName</param>
		/// <returns>The created <see cref="IMetadata"/> instance or <c>null</c> if the given QName is not supported</returns>
    IMetadata createMetadata(string localName, string namespaceUri);

		/// <summary>
		/// Creates a <see cref="IMetadata"/> instance of default type
		/// </summary>
		/// <returns>The instance</returns>
		IMetadata createMetadata();
	}
}

File: trunk/java/core/src/org/daisy/urakawa/metadata/IMetadataFactory.java
package org.daisy.urakawa.metadata;

import org.daisy.urakawa.IWithPresentation;
  
import org.daisy.urakawa.exception.MethodParameterIsEmptyStringException;
import org.daisy.urakawa.exception.MethodParameterIsNullException;
import org.daisy.urakawa.xuk.IXukAble;

/**
 * <p>
 * This is the factory that creates {@link org.daisy.urakawa.metadata.IMetadata}
 * instances.
 * </p>
 * 
 * @leafInterface see {@link org.daisy.urakawa.LeafInterface}
 * @see org.daisy.urakawa.LeafInterface
 * @stereotype OptionalLeafInterface
 * @depend - Aggregation 1 org.daisy.urakawa.IPresentation
 * @depend - Create - org.daisy.urakawa.metadata.IMetadata
 */
public interface IMetadataFactory extends IXukAble, IWithPresentation {
	/**
	 * <p>
	 * Creates a new metadata instance.
	 * </p>
	 * <p>
	 * This factory method does not take any argument and creates an object of
	 * the default type.
	 * </p>
	 * 
	 * @return cannot return null.
	 */
	public IMetadata createMetadata();

	/**
	 * <p>
	 * Creates a new metadata instance.
	 * </p>
	 * 
	 * <p>
	 * This factory method takes arguments to specify the exact type of object
	 * to create, given by the unique QName (XML Qualified Name) used in the XUK
	 * serialization format. This method can be used to generate instances of
	 * subclasses of the base object type.
	 * </p>
	 * @param xukLocalName
	 *            cannot be null, cannot be empty string.
	 * @param xukNamespaceURI
	 *            cannot be null, but can be empty string.
	 * @return can return null (in case the QName specification does not match
	 *         any supported type).
	 * @throws MethodParameterIsNullException
	 *             NULL method parameters are forbidden
	 * @throws MethodParameterIsEmptyStringException
	 *             Empty string '' method parameter is forbidden:
	 *             <b>xukLocalName</b>
	 * @tagvalue Exceptions "MethodParameterIsNull-MethodParameterIsEmptyString"
	 */
	public IMetadata createMetadata(String xukLocalName, String xukNamespaceURI)
			throws MethodParameterIsNullException,
			MethodParameterIsEmptyStringException;
}
